core/templates/domain/topic/topic-rights-backend-api.service.ts

```diff
@@ -16,162 +16,192 @@
-    };
+  }
 
-    var _sendMail = function(topicId, topicName, successCallback,
-        errorCallback) {
-      var sendMailUrl = UrlInterpolationService.interpolateUrl(
-        '/rightshandler/send_topic_publish_mail/<topic_id>', {
-          topic_id: topicId
-        });
+  private _sendMail(
+      topicId: string,
+      topicName: string,
+      successCallback: (value?: Object | PromiseLike<Object>) => void,
+      errorCallback: (reason?: string) => void): void {
+    var sendMailUrl = this.urlInterpolation.interpolateUrl(
+      '/rightshandler/send_topic_publish_mail/<topic_id>', {
+        topic_id: topicId
+      });
 
-      var putParams = {
-        topic_name: topicName
-      };
+    var putParams = {
+      topic_name: topicName
+    };
 
-      $http.put(sendMailUrl, putParams).then(function(response) {
+    this.http.put(sendMailUrl, putParams).toPromise().then(
+      (response: number) => {
         if (successCallback) {
           successCallback();
         }
-      }, function(errorResponse) {
+      }, (error) => {
         if (errorCallback) {
-          errorCallback(errorResponse.data);
+          errorCallback(error);
         }
       });
-    };
+  }
 
-    var _isCached = function(topicId) {
-      return topicRightsCache.hasOwnProperty(topicId);
-    };
+  private _isCached(
+      topicId: string): boolean {
+    return this.topicRightsCache.hasOwnProperty(topicId);
+  }
 
-    return {
-      /**
-       * Gets a topic's rights, given its ID.
-       */
-      fetchTopicRights: function(topicId) {
-        return $q(function(resolve, reject) {
-          _fetchTopicRights(topicId, resolve, reject);
-        });
-      },
-
-      /**
-       * Behaves exactly as fetchTopicRights (including callback
-       * behavior and returning a promise object), except this function will
-       * attempt to see whether the given topic rights has been
-       * cached. If it has not yet been cached, it will fetch the topic
-       * rights from the backend. If it successfully retrieves the topic
-       * rights from the backend, it will store it in the cache to avoid
-       * requests from the backend in further function calls.
-       */
-      loadTopicRights: function(topicId) {
-        return $q(function(resolve, reject) {
-          if (_isCached(topicId)) {
-            if (resolve) {
-              resolve(topicRightsCache[topicId]);
-            }
-          } else {
-            _fetchTopicRights(topicId, function(topicRights) {
-              // Save the fetched topic rights to avoid future fetches.
-              topicRightsCache[topicId] = topicRights;
-              if (resolve) {
-                resolve(topicRightsCache[topicId]);
-              }
-            }, reject);
+  /**
+   * Gets a topic's rights, given its ID.
+   */
+  fetchTopicRights(topicId: string): Promise<Object> {
+    return new Promise((resolve, reject) => {
+      this._fetchTopicRights(topicId, resolve, reject);
+    });
+  }
```

Comment:
Please use the exact type instead of `Object`.

Reply:
I'm not entirely sure what to do here, the private functions just return void

Reply:
I think the type should be equal to the type of argument with which resolve is called. In this case `TopicRightsResponseData`.

Reply:
@bansalnitish Also, I wanted to know why do we use successCallback and errorCallback. Is there a reason for it? If not,
In many places we have something like
```typescript
_funa (successCallback, errorCallback) {
  return this.http.get<>(...).then(response => {
    successCallback(response);
  }, error => {
    errorCallback(error.data);
  })
};

funa() {
  return new Promise((resolve, reject) => {
    this._funa(resolve, reject);
  });
}
```

It can be refactored to something like

```typescript
_funa() {
  return this.http.get<>(...).then(response => {
    return response;
  }, error => {
    throw new Error(error.data);
  })
};

funa() {
  return this._funa();
}
```

WDYT?

Reply:
@bansalnitish ptal!

Reply:
> I think the type should be equal to the type of argument with which resolve is called. In this case TopicRightsResponseData.

Could you please let us know why do we do this? I'm not sure. 

Also, I don't think we should go with a complete code refactor as you suggested. Adding specific type of object is fine.

Reply:
> Could you please let us know why do we do this? I'm not sure.

I'm sorry didn't get your question. Why do we do what?

Reply:
I meant passing the type which is equal to the type of argument with which resolve is called.

Reply:
It's because the Promise will return the object that is passed in `resolve` right?

Also `Object` shouldn't be used because it can cause problems later.

Consider
```typescript
function fun(): Object {
...
}

// Now doing this
let a = fun(); // The type of a would be Object

console.log(a.something) // This will give an error because something doesn't exist on Object. So, the exact type needs to be defined in this case.
```

Reply:
I think the question is: why are we passing success/failure callbacks today. It's hard to say why without asking the original author, but there is no language-specific reason we need to do this, so I think we should get rid of it. In general, all callbacks should be deleted and instead modified to return a promise.

Reply:
To be clear, I don't think we should do all the changes in this PR (or even this change, necessarily), but I do think this should be cleaned up around the code base. Let's continue discussions in #9753, and leave this comment as resolved to unblock @orthodoxparadox's work.